### PR TITLE
Case XML end to end regressions

### PR DESCRIPTION
**Pull request by ctsims**
Tests loading case transactions in/out of XML

### CODE REVIEW

**Comment by phillipm** [comment]
@damagatchi retest this please

File: tests/test/org/commcare/cases/test/CaseParseAndReadTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        Document parsed = XmlComparator.getDocumentFromStream(new ByteArrayInputStream(parsedDb));
+        Document loaded = XmlComparator.getDocumentFromStream(this.getClass().getResourceAsStream(caseDbState));
+
+        try {
+            XmlComparator.compareXmlDOMs(parsed, loaded);
+        } catch (Exception e) {
+            System.out.print(new String(parsedDb));
+
+
+            assertEquals("CaseDB output did not match expected structure(" + e.getMessage() + ")", new String(dumpStream(caseDbState)), new String(parsedDb));
```
Comment (phillipm) [inline tests/test/org/commcare/cases/test/CaseParseAndReadTest.java]: This should be moved into the body of the ```try```

File: tests/test/org/commcare/test/utils/XmlComparator.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            parser.setInput(reader);
+            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
+            document.parse(parser);
+            return document;
+        } catch(Exception e) {
+            throw new RuntimeException(e.getMessage());
+        }
+    }
+
+    public static void compareXmlDOMs(Document a, Document b) throws RuntimeException {
```
Comment (phillipm) [inline tests/test/org/commcare/test/utils/XmlComparator.java]: Please add a comment describing the matching conditions: Does order matter, are values checked, etc.

Comment (phillipm) [inline tests/test/org/commcare/test/utils/XmlComparator.java]: Shouldn't this method return a boolean and be renamed to xmlDOMsEqual? Or you leave the name the same and return -1, 0, 1, making it a comparator

File: tests/test/org/commcare/cases/test/BadCaseXMLTests.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        sandbox = MockDataUtils.getStaticStorage();
+    }
+
+    @Test
+    public void testNoCaseID() {
+        //Expected - Fail silently (TODO: Fix?)
+        MockDataUtils.parseIntoSandbox(this.getClass().getResourceAsStream("/case_create_broken_no_caseid.xml"), sandbox);
+
+        //Make sure that we didn't make a case entry for the bad case though
+        //assertEquals("Case XML with no id should not have created a case record", sandbox.getCaseStorage().getNumRecords(), 0);
```
Comment (phillipm) [inline tests/test/org/commcare/cases/test/BadCaseXMLTests.java]: I've created a ticket for this: 
http://manage.dimagi.com/default.asp?178270

**Comment by phillipm** [comment]
@ctsims Good to go after addressing my inline comments.

File: tests/test/org/commcare/cases/test/CaseParseAndReadTest.java (hunk 1 above)
Comment (ctsims) [inline tests/test/org/commcare/cases/test/CaseParseAndReadTest.java]: Hey,

So this is actually basically an "assertFalse" that happens to outline the correct sections of the xml more often than not.

The two strings aren't actually guaranteed to be the same even _if_ the XML is semantically equivalent, but if they _are_ different, the strings definitely will be, and the areas where they are different get highlighted quite nicely in my tests

File: tests/test/org/commcare/test/utils/XmlComparator.java (hunk 2 above)
Comment (ctsims) [inline tests/test/org/commcare/test/utils/XmlComparator.java]: agreed. Will document

**Comment by ctsims** [comment]
@phillipm Addressed comments and fixed some linting

**Comment by phillipm** [comment]
:+1:

**Pull request merged**